File: python/widgets/gauge.py
from PySide6.QtWidgets import QWidget
from PySide6.QtGui import QPainter, QPen, QFont
from PySide6.QtCore import Qt, QRectF, QPointF
import theme

class Gauge(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.value = 0
        self.setMinimumSize(220, 220)

    def set_value(self, val):
        self.value = val
        self.update()

    def paintEvent(self, event):
        p = QPainter(self)
        p.setRenderHint(QPainter.Antialiasing)
        cx, cy = self.width()/2, self.height()/2
        p.translate(cx, cy)
        
        # Ticks
        p.setPen(QPen(theme.ACTIVE["secondary"], 1))
        for i in range(0, 360, 10):
            p.save()
            p.rotate(i)
            p.drawLine(QPointF(90, 0), QPointF(96, 0))
            p.restore()

        # Active Arc
        rect = QRectF(-85, -85, 170, 170)
        p.setPen(QPen(theme.ACTIVE["primary"], 6))
        span = int((self.value / 100.0) * 360)
        # PySide6 drawArc uses 1/16th degrees
        p.drawArc(rect, 90 * 16, -span * 16)
        
        p.setPen(Qt.white)
        p.setFont(QFont("Consolas", 28, QFont.Bold))
        p.drawText(rect, Qt.AlignCenter, f"{int(self.value)}")

File: src/GUI/widgets/gauge.cpp
// Location: src/GUI/widgets/gauge.cpp
#include "gauge.h"
#include <QFontMetrics>
#include <QPaintEvent>
#include <QPainter>
Gauge::Gauge(QWidget *parent) : QWidget(parent) { setMinimumSize(100, 100); }
void Gauge::setValue(float value) {
    if (value < m_min) value = m_min;
    if (value > m_max) value = m_max;
    if (m_value == value) return;
    m_value = value;
    update();
}
void Gauge::setRange(float minValue, float maxValue) {
    m_min = minValue;
    m_max = maxValue;
    update();
}
void Gauge::setLabel(const QString &text) {
    m_label = text;
    update();
}
void Gauge::paintEvent(QPaintEvent *) {
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);
    int w = width();
    int h = height();
    int size = qMin(w, h);
    p.translate(w / 2, h / 2);
    p.scale(size / 200.0, size / 200.0);
    // Background
    p.setBrush(QColor(40, 40, 40));
    p.drawEllipse(-90, -90, 180, 180);
    // Value Arc
    float normalized = (m_value - m_min) / (m_max - m_min);
    p.setPen(QPen(Qt::cyan, 10));
    p.drawArc(-80, -80, 160, 160, -30 * 16, (240 * normalized) * 16);
    // Label
    p.setPen(Qt::white);
    p.drawText(-40, 40, 80, 20, Qt::AlignCenter, m_label);
}